### PR TITLE
Governance v2 (Kusama only)

### PR DESCRIPTION
**Pull request by gavofyork**
Prior to merge:
- [x] Check UI integration
  - [x] https://github.com/paritytech/substrate/issues/12272
- [x] Check Kusama release & governance schedule
- [ ] Final sanity check on the graphs (including dumb stuff like units)
- [x] Consider plan for seeding PF membership

TODO:

- [x] Integrate multi-level treasury spending (depends on https://github.com/paritytech/substrate/issues/11116 / https://github.com/paritytech/substrate/pull/11124)
- [x] Introduce and integrate Whilelist pallet (https://github.com/paritytech/substrate/pull/10159) to use Fellowship
- [x] Introduce new curves for turnout/approval (https://github.com/paritytech/substrate/pull/11124)
- [x] Introduce a simple ranked membership for the Fellowship (https://github.com/paritytech/substrate/pull/11474)

### CODE REVIEW

File: runtime/kusama/src/gov_config.rs
Comment (xlc) [inline runtime/kusama/src/gov_config.rs:106]: Does it make sense to have `Spender { max: Balance }`?

Comment (gavofyork) [reply runtime/kusama/src/gov_config.rs]: Yes. That's done now.

**Review by ggwpez — COMMENTED** [review]
I tried to run the benchmarks locally, and all but conviction voting errored out.  
Lets see what the bot says.

**Comment by ggwpez** [comment]
/cmd queue -c bench-bot $ runtime kusama-dev pallet_conviction_voting

File: runtime/kusama/src/governance/origins.rs
Comment (xlc) [inline runtime/kusama/src/governance/origins.rs:33]:
```suggestion
	#[derive(PartialEq, Eq, Clone, MaxEncodedLen, Encode, Decode, TypeInfo, RuntimeDebug)]
```

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_conviction_voting

File: runtime/kusama/src/governance/old.rs
Comment (shawntabrizi) [inline runtime/kusama/src/governance/old.rs:3]: should sanity check this file and configuration one last time before we merge, to make sure that master config doesn't change.

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

File: runtime/kusama/constants/src/lib.rs
Comment (KiChjang) [inline runtime/kusama/constants/src/lib.rs:30]: I would have assumed that a quid is equivalent to a unit, how is this not the case?

Comment (shawntabrizi) [reply runtime/kusama/constants/src/lib.rs:30]: Why do you have that assumption?

I don't think we would have two ways to write an amount if they would be the same.

Unit represents "1 KSM"

Whereas "QUID" or "DOLLAR" is some abstract, **non-fiat** representation of what we use to base fees and other storage deposits off of.

When we adjust fees, for example make them cheaper, we simply adjust what `cents` are, and all fees should be adjusted accordingly.

File: runtime/kusama/src/governance/mod.rs
Comment (muharem) [inline runtime/kusama/src/governance/mod.rs:72]: Does it mean that `Fellows` (from 3rd rank and higher) can let a `WhitelistedCaller` to execute any call as `Root`?
https://github.com/paritytech/substrate/blob/97ae6be11b0132224a05634c508417f048894670/frame/whitelist/src/lib.rs#L224

File: runtime/kusama/constants/src/lib.rs
Comment (KiChjang) [reply runtime/kusama/constants/src/lib.rs:30]: Okay the justification that you just told me was non-trivial, sounds like something that we should mention as a comment.

Comment (shawntabrizi) [reply runtime/kusama/constants/src/lib.rs:30]: Happy to resolve that as a follow up issue. lets not bring that into this big pr

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_ranked_collective

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_whitelist

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_ranked_collective

File: runtime/kusama/src/governance/mod.rs
Comment (shawntabrizi) [reply runtime/kusama/src/governance/mod.rs:72]: No, it means that Root can call functions where we check for `WhistelistOrigin`: https://github.com/paritytech/substrate/blob/97ae6be11b0132224a05634c508417f048894670/frame/whitelist/src/lib.rs#L122

**Review by shawntabrizi — APPROVED** [review]
I want to make a tool which graphs all of these different curves and gives a visual of how each level works, but generally, i think it is good to go.

There are questions above which can only be answered by @gavofyork, mostly on intentionality of values

File: runtime/kusama/src/governance/fellowship.rs
Comment (ggwpez) [inline runtime/kusama/src/governance/fellowship.rs:270]: Can we remove this if we use the correct `success_origin` everywhere?

**Review by ggwpez — APPROVED** [review]

File: runtime/kusama/src/lib.rs
Comment (ggwpez) [inline runtime/kusama/src/lib.rs:1486]: This results in the pallet being listed twice with `benchmark pallet --list --chain kusama-dev`.  
I think we can append the instance to the pallet name in the benchmarking CLI.  
What do you think?

Comment (shawntabrizi) [reply runtime/kusama/src/lib.rs:1486]: yes, we need some better solution to handle instances. not sure exactly what that is, but it should show up in the list, and be targeted with the appropriate command, and have a unique output

**Comment by shawntabrizi** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

File: runtime/kusama/constants/src/lib.rs
Comment (ggwpez) [reply runtime/kusama/constants/src/lib.rs:30]: https://github.com/paritytech/polkadot/issues/5912

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_ranked_collective

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_whitelist

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_conviction_voting

**Comment by ggwpez** [comment]
Looks like the bot is missing dependencies. I posted it in devops.

**Comment by ggwpez** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

**Comment by ggwpez** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_whitelist

**Comment by ggwpez** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_conviction_voting

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_whitelist

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_fellowship

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_referenda

**Comment by KiChjang** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_referenda

**Comment by ggwpez** [comment]
The benchmarking does not deal with multiple instances which is why it fails, see https://github.com/paritytech/substrate/issues/12080.

**Comment by ggwpez** [comment]
The bench bot should now handle multi-instance pallets the same way it handles single-instance pallets.  
Such that `pallet-referenda` would produce two weight files when called with the standard command.

File: runtime/kusama/src/governance/fellowship.rs
Comment (4meta5) [inline runtime/kusama/src/governance/fellowship.rs:309]:
```suggestion
	type Slash = Treasury;
```

File: runtime/kusama/src/lib.rs
Comment (gavofyork) [reply runtime/kusama/src/lib.rs:638]: This one in particular won't be used any more since we have `SpendOrigin` to take over from it. Others are likely for components which are also defunct now. If there's one in particular you're concerned about, best to specify it explicitly.

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_referenda

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_fellowship

**Comment by ggwpez** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_referenda

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_fellowship

File: Cargo.lock
Comment (ordian) [inline Cargo.lock:126]: New version of arbitrary bumped MSRV to 1.63. Our CI doesn't have it yet.

It looks like you did `cargo update` which we strongly discourage in favor of reviewing individual dependabot updates that are actually reviewed.

I'd suggest reverting all of those changes 
```
cargo checkout master Cargo.lock
```

Comment (ggwpez) [reply Cargo.lock:126]: Looks like something in fe65488d6092130040d7a9008663c11b8ed7ca2e1 went wrong while merging.

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_fellowship

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda_referenda

**Comment by gavofyork** [comment]
/cmd queue -v RUST_LOG=debug -c bench-bot $ runtime kusama-dev pallet_referenda

**Review by ordian — APPROVED** [review]
This is just an approval without a review to unblock the merge.

**Comment by ordian** [comment]
kusama benchmarks doesn't work: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/1905517
```
2022-10-02 10:03:00 panicked at 'Expected Ok(_). Got Err(
    Module(
        ModuleError {
            index: 22,
            error: [
                8,
                0,
                0,
                0,
            ],
            message: Some(
                "NoPermission",
            ),
        },
    ),
)', /usr/local/cargo/git/checkouts/substrate-7e08433d4c370a21/952030c/frame/ranked-collective/src/benchmarking.rs:42:9   
```

File: runtime/kusama/src/governance/fellowship.rs
Comment (gavofyork) [reply runtime/kusama/src/governance/fellowship.rs:119]:
```suggestion
					min_enactment_period: 1 * MINUTES,
```

File: runtime/kusama/src/governance/origins.rs
Comment (gavofyork) [reply runtime/kusama/src/governance/origins.rs:171]: the numbers are all pretty arbitrary - we can tweak over time through further upgrades.

File: runtime/kusama/src/lib.rs
Comment (gavofyork) [reply runtime/kusama/src/lib.rs:242]: legit comment documenting a concern which is not immediate actionable.

File: runtime/kusama/src/governance/tracks.rs
Comment (gavofyork) [reply runtime/kusama/src/governance/tracks.rs:98]: possibly.

**Comment by ggwpez** [comment]
> kusama benchmarks job doesn't work: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/1905517

Fixed in 93138b7b30a050ad6a84a58aa14d82a0be1f3976.

**Review by andresilva — APPROVED** [review]

File: runtime/kusama/src/lib.rs
Comment (gavofyork) [reply runtime/kusama/src/lib.rs:908]: possibly, but should be done in a separate PR if so.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/polkadot-release-analysis/1026/2

**Pull request merged**
